src/poseidon/traits.rs

```diff
@@ -0,0 +1,416 @@
+use crate::poseidon::grain_lfsr::PoseidonGrainLFSR;
+use crate::poseidon::PoseidonParameters;
+use ark_ff::{fields::models::*, FpParameters, PrimeField};
+use ark_std::{vec, vec::Vec};
+
+/// A trait for default Poseidon parameters associated with a prime field
+pub trait PoseidonDefaultParameters: FpParameters {
+    /// An array of the parameters optimized for constraints
+    /// (rate, alpha, full_rounds, partial_rounds, skip_matrices)
+    /// for rate = 2, 3, 4, 5, 6, 7, 8
+    ///
+    /// Here, `skip_matrices` denote how many matrices to skip before
+    /// finding one that satisfy all the requirements.
+    const PARAMS_OPT_FOR_CONSTRAINTS: [[usize; 5]; 7];
+
+    /// An array of the parameters optimized for weights
+    /// (rate, alpha, full_rounds, partial_rounds, skip_matrices)
+    /// for rate = 2, 3, 4, 5, 6, 7, 8
+    const PARAMS_OPT_FOR_WEIGHTS: [[usize; 5]; 7];
+}
+
+/// A field with Poseidon parameters associated
+pub trait PoseidonDefaultParametersField: PrimeField {
+    /// Obtain the default Poseidon parameters for this rate and for this prime field,
+    /// with a specific optimization goal.
+    fn get_default_parameters(
+        rate: usize,
+        optimized_for_constraints: bool,
+    ) -> Option<PoseidonParameters<Self>>;
+
+    /// Internal function that uses the `PoseidonDefaultParameters` to compute the Poseidon parameters.
+    fn get_default_parameters_internal<P: PoseidonDefaultParameters>(
+        rate: usize,
+        optimized_for_constraints: bool,
+    ) -> Option<PoseidonParameters<Self>> {
+        let params_set = if !optimized_for_constraints {
+            P::PARAMS_OPT_FOR_CONSTRAINTS
+        } else {
+            P::PARAMS_OPT_FOR_WEIGHTS
+        };
+
+        for param in params_set.iter() {
+            if param[0] == rate {
+                let (ark, mds) = Self::find_ark_and_mds(
+                    P::MODULUS_BITS as u64,
+                    rate,
+                    param[2] as u64,
+                    param[3] as u64,
+                    param[4] as u64,
+                );
+
+                return Some(PoseidonParameters {
+                    full_rounds: param[2],
+                    partial_rounds: param[3],
+                    alpha: param[1] as u64,
+                    ark: ark,
+                    mds: mds,
+                    rate: param[0],
+                    capacity: 1,
+                });
+            }
+        }
+
+        None
+    }
+
+    /// Internal function that computes the ark and mds from the Poseidon Grain LFSR.
+    fn find_ark_and_mds(
```

Comment:
Same here: Is it possible to move this function to `PoseidonDefaultParameters` and takes generic type `<F: PoseidonDefaultParametersField>`? By doing so, those internal function will not be exposed to user

Reply:
Great points. In a fork on AleoHQ, I moved these functions out of the traits. I will likely do the same here.